File: src/owncloud/cron.py
from crontab import CronTab
from subprocess import check_output
from syncloud_app import logger


class OwncloudCron:

    def __init__(self, cron_cmd, cron_user):
        self.cron_cmd = cron_cmd
        self.cron_user = cron_user
        self.cron = CronTab(user=self.cron_user)
        self.log = logger.get_logger('owncloud.cron')

    def remove(self):
        print("remove crontab task")

        for job in self.cron.find_command(self.cron_user):
            self.cron.remove(job)
        self.cron.write()

    def create(self):
        print("create crontab task")
        ci_job = self.cron.new(command=self.cron_cmd)
        ci_job.setall('*/15 * * * *')
        self.cron.write()

    def run(self):
        self.log.info("running: {0}".format(self.cron_cmd))
        self.log.info(check_output('sudo -H -u {0} {1}'.format(self.cron_user, self.cron_cmd), shell=True))

File: cli/installer/cron.go
package installer

import (
	"fmt"
	"os/exec"
	"strings"

	"go.uber.org/zap"
)

type Cron struct {
	user     string
	command  string
	executor *Executor
	logger   *zap.Logger
}

func NewCron(user string, executor *Executor, logger *zap.Logger) *Cron {
	return &Cron{
		user:     user,
		command:  fmt.Sprintf("/usr/bin/snap run %s.cron", App),
		executor: executor,
		logger:   logger,
	}
}

func (c *Cron) Remove() error {
	c.logger.Info("removing crontab task")
	out, err := exec.Command("crontab", "-u", c.user, "-l").CombinedOutput()
	if err != nil {
		// no crontab yet for this user is not an error
		return nil
	}
	var kept []string
	for _, line := range strings.Split(string(out), "\n") {
		if !strings.Contains(line, c.command) {
			kept = append(kept, line)
		}
	}
	return c.write(strings.Join(kept, "\n"))
}

func (c *Cron) Create() error {
	c.logger.Info("creating crontab task")
	existing := ""
	out, err := exec.Command("crontab", "-u", c.user, "-l").CombinedOutput()
	if err == nil {
		existing = strings.TrimRight(string(out), "\n")
	}
	entry := fmt.Sprintf("*/15 * * * * %s", c.command)
	if existing == "" {
		return c.write(entry + "\n")
	}
	return c.write(existing + "\n" + entry + "\n")
}

func (c *Cron) Run() error {
	c.logger.Info("running cron", zap.String("cmd", c.command))
	_, err := c.executor.Run("/usr/bin/snap", "run", App+".cron")
	return err
}

func (c *Cron) write(content string) error {
	cmd := exec.Command("crontab", "-u", c.user, "-")
	cmd.Stdin = strings.NewReader(content)
	out, err := cmd.CombinedOutput()
	if err != nil {
		return fmt.Errorf("crontab write failed: %w: %s", err, string(out))
	}
	return nil
}
